src/muse/timeslices.py

```diff
@@ -105,29 +76,191 @@ def setup_module(settings: Union[str, Mapping]):
     TIMESLICE = read_timeslices(settings)
 
 
-def broadcast_timeslice(x, ts=None):
+def broadcast_timeslice(
+    data: DataArray, ts: Optional[DataArray] = None, level: Optional[str] = None
+) -> DataArray:
+    """Convert a non-timesliced array to a timesliced array by broadcasting.
+
+    If x is already timesliced in the appropriate scheme, it will be returned unchanged.
+
+    Args:
+        data: Array to broadcast.
+        ts: Dataarray with timeslice lengths. If None, defaults to the global timeslice.
+        level: Level to broadcast to. If None, use the finest level of ts.
+
+    """
     from xarray import Coordinates
 
     if ts is None:
         ts = TIMESLICE
 
-    # If x already has timeslices, check that it is matches the reference timeslice.
-    if "timeslice" in x.dims:
-        if x.timeslice.reset_coords(drop=True).equals(ts.timeslice):
-            return x
+    if level is not None:
+        ts = compress_timeslice(ts, ts=ts, level=level, operation="sum")
+
+    # If x already has timeslices, check that it matches the reference timeslice.
+    if "timeslice" in data.dims:
+        if data.timeslice.reset_coords(drop=True).equals(ts.timeslice):
+            return data
         raise ValueError("x has incompatible timeslicing.")
 
     mindex_coords = Coordinates.from_pandas_multiindex(ts.timeslice, "timeslice")
-    extensive = x.expand_dims(timeslice=ts["timeslice"]).assign_coords(mindex_coords)
-    return extensive
+    broadcasted = data.expand_dims(timeslice=ts["timeslice"]).assign_coords(
+        mindex_coords
+    )
+    return broadcasted
+
+
+def distribute_timeslice(
+    data: DataArray, ts: Optional[DataArray] = None, level=None
+) -> DataArray:
+    """Convert a non-timesliced array to a timesliced array by distribution.
+
+    If x is already timesliced in the appropriate scheme, it will be returned unchanged.
+
+    Args:
+        data: Array to distribute.
+        ts: Dataarray with timeslice lengths. If None, defaults to the global timeslice.
+        level: Level to distribute to. If None, use the finest level of ts.
+
+    """
+    if ts is None:
+        ts = TIMESLICE
+
+    if level is not None:
+        ts = compress_timeslice(ts, ts=ts, level=level, operation="sum")
+
+    # If x already has timeslices, check that it matches the reference timeslice.
+    if "timeslice" in data.dims:
+        if data.timeslice.reset_coords(drop=True).equals(ts.timeslice):
+            return data
+        raise ValueError("x has incompatible timeslicing.")
+
+    broadcasted = broadcast_timeslice(data, ts=ts)
+    timeslice_fractions = ts / broadcast_timeslice(ts.sum(), ts=ts)
+    return broadcasted * timeslice_fractions
```

Comment:
Just to make sure I get this right:
- First we broadcast the input array over the target timeslice - same value in all timeslices if it had no timeslice info, to start with
- Then, we multiply by the timeslice fractions to ensure that the data is proportional to the timeslice fraction.

Reply:
Correct